write_hooks.go

```diff
@@ -0,0 +1,223 @@
+package qfs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+)
+
+// WriteHook is a function that's called when a given path has been
+// written to the content addressed filesystem
+type WriteHook func(ctx context.Context, f File, added map[string]string) (io.Reader, error)
+
+// hookFile configures a callback function to be executed on a saved
+// file, at a specific point in the merkelization process
+type hookFile struct {
+	// file for delayed hook calls
+	File
+	// once mutex for callback execution
+	once sync.Once
+	// slice of pre-merkelized paths that need to be saved before the hook
+	// can be called
+	requiredPaths []string
+	// function to call
+	callback WriteHook
+}
+
+// Assert hookFile implements HookFile at compile time
+var _ WriteHookFile = (*hookFile)(nil)
+
+// WriteHookFile is a file that can hook into the merkelization process, affecting
+// contents as contents are being rendered immutable
+type WriteHookFile interface {
+	File
+	RequiredPaths() []string
+	HasRequiredPaths(paths map[string]string) bool
+	CallAndAdd(ctx context.Context, adder Adder, added map[string]string) error
+}
+
+// NewWriteHookFile wraps a File with a hook & set of sibling / child dependencies
+func NewWriteHookFile(file File, cb WriteHook, requiredPaths ...string) File {
+	if file.IsDirectory() {
+		panic("cannot create a WriteHookFile with a Directory")
+	}
+
+	return &hookFile{
+		File:          file,
+		requiredPaths: requiredPaths,
+		callback:      cb,
+	}
+}
+
+func (h *hookFile) RequiredPaths() []string {
+	return h.requiredPaths
+}
+
+func (h *hookFile) HasRequiredPaths(merkelizedPaths map[string]string) bool {
+	for _, p := range h.requiredPaths {
+		if _, ok := merkelizedPaths[p]; !ok {
+			log.Debugf("hook %q can't fire. waiting for %s", h.FullPath(), p)
+			return false
+		}
+	}
+	return true
+}
+
+func (h *hookFile) CallAndAdd(ctx context.Context, adder Adder, merkelizedPaths map[string]string) (err error) {
+	h.once.Do(func() {
+		log.Debugf("calling hook path=%s merkelized=%#v", h.FullPath(), merkelizedPaths)
+		var r io.Reader
+		r, err = h.callback(ctx, h.File, merkelizedPaths)
+		if err != nil {
+			return
+		}
+		if err = adder.AddFile(ctx, NewMemfileReader(h.FullPath(), r)); err != nil {
+			return
+		}
+	})
+
+	return err
```

Comment:
Took me a number of times reading this function to figure out if `err` can ever be non-nil here. I don't think it can? I think it would be clearer if the named return value was unnamed instead `(err error)` -> `error`, and `var r io.Reader` was removed, in favor of `r, err := h.callback...`, with `return nil` to finish the function.

Reply:
`err` can absolutely be non nil here.

The strange semantics here are thanks to the `sync.Once.Do` call, which requires a function argument with the signature `func()`. Because we don't get an error return, the error variable must be declared outside the scope at the time same time as the function arg declaration. and set within the inner scope. The first time this function is called any error in execution will set the error variable, which is returned on line 81.